ir/src/main/scala/org/scalajs/ir/ScalaJSVersions.scala

```diff
@@ -12,32 +12,94 @@
-   *  This should be either of:
-   *  - a prior release version (i.e. "0.5.0", *not* "0.5.0-SNAPSHOT")
-   *  - `current`
-   */
-  val binaryEmitted: String = current
-
-  /** Versions whose binary files we can support (used by deserializer) */
-  val binarySupported: Set[String] = {
-    Set(binaryEmitted)
+  current = "1.0.0-SNAPSHOT",
+  /** Version of binary IR emitted by this version of Scala.js. */
+  binaryEmitted = "1.0-SNAPSHOT"
+)
+
+/** Helper class to allow for testing of logic. */
+class VersionChecks private[ir] (
+    final val current: String,
+    final val binaryEmitted: String
+) {
+  import VersionChecks._
+
+  checkConsistent(current, binaryEmitted)
+
+  private val (binaryMajor, binaryMinor, binaryPreRelease) = parseBinary(binaryEmitted)
+
+  private val knownSupportedBinary = {
+    val m = new ConcurrentHashMap[String, Unit]()
+    m.put(binaryEmitted, ())
+    m
+  }
+
+  /** Check we can support this binary version (used by deserializer) */
+  final def checkSupported(version: String): Unit = {
+    if (!knownSupportedBinary.containsKey(version)) {
+      val (major, minor, preRelease) = parseBinary(version)
+      val supported = (
+          // the exact pre-release version is supported via knownSupportedBinary
+          preRelease.isEmpty &&
+          major == binaryMajor &&
+          minor <= binaryMinor &&
+          (binaryPreRelease.isEmpty || minor < binaryMinor)
+      )
+
+      if (supported) {
+        knownSupportedBinary.put(version, ())
+      } else {
+        throw new IRVersionNotSupportedException(version, binaryEmitted,
+            s"This version ($version) of Scala.js IR is not supported. " +
+            s"Supported versions are up to $binaryEmitted")
+      }
+    }
   }
+}
 
-  // Just to be extra safe
-  assert(binarySupported contains binaryEmitted)
+private object VersionChecks {
+  private val binaryRE = """^([0-9]+)\.([0-9]+)(-.*)?$""".r
+
+  private def parseBinary(v: String): (Int, Int, Option[String]) = {
+    val m = mustMatch(binaryRE, v)
+    (m.group(1).toInt, m.group(2).toInt, Option(m.group(3)))
+  }
 
+  private def parseFull(v: String): (Int, Int, Int, Option[String]) = {
+    val fullRE = """^([0-9]+)\.([0-9]+)\.([0-9]+)(-.*)?$""".r
+    val m = mustMatch(fullRE, v)
+    (m.group(1).toInt, m.group(2).toInt, m.group(3).toInt, Option(m.group(4)))
+  }
+
+  private def mustMatch(re: Regex, v: String): Regex.Match = {
+    re.findFirstMatchIn(v).getOrElse(
+        throw new IllegalArgumentException("malformed version: " + v))
+  }
+
+  private def checkConsistent(current: String, binary: String) = {
+    val (binaryMajor, binaryMinor, binaryPreRelease) = parseBinary(binary)
+    val (currentMajor, currentMinor, currentPatch, currentPreRelease) = parseFull(current)
+
+    require(currentMajor == binaryMajor, "major(current) != major(binaryEmitted)")
+
+    require(currentMinor >= binaryMinor, "minor(current) < minor(binaryEmitted)")
+
+    require(currentPreRelease.isEmpty ||
```

Comment:
Start the actual argument on the next line, since it spans multiple lines:
```scala
    require(
        currentPreRelease.isEmpty ||
        ...
```